### ISSUE TITLE
Program Cache Collision in Batch Norm and Permute Operations

### ISSUE THREAD

**Issue by ndivnicTT** [tenstorrent/tt-metal#46018]
# Program Cache Collision in Batch Norm and Permute Operations

## Summary

The fused group 2 CI failure looked like an SFPU regression because it surfaced in batch normalization mixed-precision coverage after an SFPU layer refactor. The root cause was program cache key reuse across operations whose generated programs depend on tensor structure that was not fully represented in their program hashes.

Two related cache collision paths were found:

- `BatchNormOperation` and `RunningStatistics` hashed only a limited subset of tensor metadata.
- `PermuteDeviceOperation`, reached through `ttnn.mean` inside training batch norm, relied on the default operation hash and did not explicitly include descriptor-shaping inputs.

Both collisions could reuse a cached program built for a different tensor shape/layout/memory configuration. In the failing sequence this produced invalid runtime argument access and, in longer runs, a segmentation fault.

## Observed Failure

The original failing CI split was:

```bash
pytest --timeout 300 tests/ttnn/unit_tests/operations/fused -xv \
  --splits 2 --group 2 --splitting-algorithm least_duration \
  -m 'not disable_fast_runtime_mode'
```

The visible error was:

```text
TT_FATAL: Index 0 is larger than runtime args size 0
```

followed by process termination with exit code `139` in some full-split runs.

The failure was order-dependent. Individual batch norm tests could pass, while the full split or a smaller two-test sequence reproduced the crash because the program cache already contained an incompatible program.

## Diagnosis

The initial crash was reproduced in `tests/ttnn/unit_tests/operations/fused/test_batch_norm.py`, then narrowed to a program cache collision in batch norm-related operations.

After fixing the direct batch norm cache key, the full split still found a second history-dependent failure. Further minimization showed that training batch norm calls `ttnn.mean`, which internally reaches `PermuteDeviceOperation`. That operation also needed an explicit program hash because its selected program and runtime descriptors depend on the permutation dimensions and tensor structure.

This means the issue was correlated with SFPU coverage, but not caused by SFPU math/typecast itself. The SFPU-heavy batch norm tests exposed insufficient program cache keys.

### Batch Norm

Updated `BatchNormOperation::compute_program_hash` in:

- `ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp`

The hash now includes full structural metadata for all tensors that influence program generation:

- logical shape
- padded shape
- physical volume
- dtype
- layout
- memory config
- tile height and width
- presence and metadata of optional tensors

This replaced the previous hash, which only used operation attributes plus limited dtype and memory config data.

### Running Statistics

Added and implemented `RunningStatistics::compute_program_hash` in:

- `ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_device_operation.hpp`
- `ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_device_operation.cpp`

The new hash mirrors the batch norm tensor-structure hashing for:

- `batch_mean`
- `batch_var`
- optional `running_mean`
- optional `running_var`

### Permute

`PermuteDeviceOperation::compute_program_hash` is present in:

- `ttnn/cpp/ttnn/operations/data_movement/permute/device/permute_device_operation.hpp`
- `ttnn/cpp/ttnn/operations/data_movement/permute/device/permute_device_operation.cpp`

The branch implementation explicitly includes:

- permutation dimensions
- output memory config
- pad value
- selected program factory index
- input tensor spec and padded shape
- computed output spec and padded shape

This prevents program cache hits between permutes that have the same high-level operation type but require different generated programs or runtime descriptors.

A later Blackhole release build failure was caused by an accidental duplicate declaration and duplicate definition of `PermuteDeviceOperation::compute_program_hash`. The fix for that compile failure is to keep the branch implementation above and remove the duplicate earlier declaration/definition.

## Validation

The following targeted and full-split validations passed after rebuilding `ttnn` and ensuring Python loaded the rebuilt `_ttnn.so`:

```bash
pytest --timeout 300 tests/ttnn/unit_tests/operations/fused/test_batch_norm.py -xv
```

```bash
pytest --timeout 300 tests/ttnn/unit_tests/operations/fused -xv \
  --splits 2 --group 2 --splitting-algorithm least_duration \
  -m 'not disable_fast_runtime_mode'
```

Full split result:

- `2267 selected`
- exit code `0`
- elapsed time approximately `15m53s`

The Blackhole release compile failure from `https://github.com/tenstorrent/tt-metal/actions/runs/26898692224/job/79344862430` was also reproduced from the logs:

```text
permute_device_operation.hpp:102:31: error: class member cannot be redeclared
permute_device_operation.cpp:127:44: error: redefinition of 'compute_program_hash'
```

After removing the duplicate permute declaration/definition, local validation passed:

```bash
ninja -C build ttnn
```

No IDE linter errors were reported for the edited files.

## Issue Notes

This looks like a broader class of program cache correctness bugs: operations with custom program factories should ensure their cache key includes every attribute and tensor property that affects generated kernels, circular buffers, runtime arguments, sharding, memory layout, or output descriptors.

Potential follow-up work:

- Audit other device operations that rely on default hashing while selecting different program factories or descriptor shapes.
- Add targeted cache-collision tests that run the same op type with different tensor structures in sequence with program cache enabled.
- Consider sharing a helper for hashing tensor structure consistently across operations.
